Organization<>Team<>User structure (closes #1540)

Twixes opened the pull request:
## Changes

Taking over from #1461, this adds an Organization<>Team<>User structure, with team creation being a premium feature.

## Checklist

- [x] All querysets/queries filter by Team (if this PR affects any querysets/queries)
- [ ] Backend tests (if this PR affects the backend)
- [ ] Cypress E2E tests (if this PR affects the front and/or backend)

Algogator committed:
Backend changes to implement #1461

Algogator committed:
Added the missing migration files

Algogator committed:
Fixes Typecheck errors

Algogator committed:
Refactor request.user.team_set.get() to use request.user.team

Algogator committed:
Updated user patch method to change current_team on team id instead of name

Algogator committed:
Merge branch 'master' of https://github.com/PostHog/posthog into teams

Algogator committed:
Merged migration file

Algogator committed:
Changes team property to return first item in queryset

Algogator committed:
Fixes failing tests

Algogator committed:
Changed User api to return the ids of the teams they are currently part of

Commit:
Frontend for changing teams

Twixes committed:
Merge branch 'master' into pr/1468

Twixes committed:
Update and simplify migrations

Twixes committed:
Improve team and user buttons

Twixes committed:
Make team changing backend more logical

Twixes committed:
Improve current_team mechanics

Twixes committed:
Update test_team.py

Twixes committed:
Fix Team.objects.create_with_data

Twixes committed:
Update migration

Twixes committed:
Update tests

Twixes committed:
Make setup_review more convenient

Twixes committed:
Add Organization and OrganizationMembership

Twixes committed:
Replace is_admin with level

Twixes committed:
Extend API

Twixes committed:
Update team.py

Twixes committed:
Improve modeling

Twixes committed:
Improve handling of new mechanics

Twixes committed:
Add proper migration

Twixes committed:
Remove _ensure_organization_and_team

Twixes committed:
Update 0084_org_team_user.py

Twixes committed:
Improve user, org and team creation

Twixes committed:
Make MembershipLevel more flexible for the future

Twixes committed:
Add member deletion

Twixes committed:
Fix naive datetime warnings

Twixes committed:
Update setup_review.py

Twixes committed:
Update API route

Twixes committed:
Make PersonalAPIKey changes

Twixes committed:
Update models and migrations, fix typing

Twixes committed:
Merge branch 'master' into teams

Twixes committed:
Fix typing

Twixes committed:
Use MAC-less UUID v1 instead of v4 for better performance

Twixes committed:
Add abstract UUIDModel

Twixes committed:
Update utils.py

Twixes committed:
Update utils.py

Twixes committed:
Fix multi/unicast bit

Twixes committed:
Update APIs, frontend and tests

Twixes committed:
Update pull_request_template.md

Twixes committed:
Merge branch 'master' into teams

Twixes commented:
I believe this can be considered done. Though a few things to get everything wrapped up nicely:
* I created an Organization > Settings page thinking that there'll be some setting there… but it's still empty, as I'm not really sure what would make sense to be moved from Team to Organization (Members, Invites, and Licenses have their own pages, so they are out of consideration). Let me know your thoughts, else the page can be removed for now (though it will surely come useful in the future).
* I absolutely **cannot figure out why org member removal refuses to work**. Looks like `member_to_delete = self.get_object()` returns something different than… the member to delete… but I'm not sure why. Please take a look, it may be completely trivial!
* TODO: At the moment only org founder can be one of its admins. I'll implement changing member levels in another PR, as this is already pretty big. Not a big deal, as nobody can remove an admin and they can't leave, so this is just similar to an owner concept for now.
* Speaking of an owner concept, I decided not to implement that in addition to "member" (level 1) and "administrator" (level 8 – the gap is if we ever want something in-between), current structure should be sufficient.
* TODO: We can now nicely replace `is_demo` Hogflix data (along with the "Delete demo data" button), by simply making Hogflix data a separate team – removable in its entirety any time.

timgl commented:
Nicely done, I'll review this but already looks like some failing tests

timgl commented on posthog/migrations/0084_org_team_user.py:
```diff
@@ -0,0 +1,175 @@
+    OrganizationMembership = apps.get_model("posthog", "OrganizationMembership")
+    for user in User.objects.all():
+        team = user.teams_deprecated_relationship.get()
+        deterministic_derived_uuid = team.deterministic_derived_uuid
```
This line causes running migrations locally to fail with "AttributeError: 'Team' object has no attribute 'deterministic_derived_uuid'"

timgl commented:
I'm unable to test this due to the migration issue. In general, I think this pull request is too big for us to safely merge. Is there a way we could split this PR in smaller chunks without too much work?

Twixes commented:
Not really, this is all very much intertwined.

Twixes committed:
Fix comment

Twixes committed:
Fix migration error

Twixes committed:
Compress migrations

Twixes committed:
Updates with minimal renaming

Twixes committed:
More updates

Twixes committed:
Make further updates

Twixes commented:
For posterity: these changes will be merged in a more gradual way, hence this is a draft.

Twixes committed:
Merge branch 'master' into teams

Twixes committed:
Update test_team_user.py

Twixes committed:
Fix issues

Twixes committed:
Add migration

Twixes committed:
Satisfy mypy

Twixes committed:
Remove Signup redirect on logged in

Twixes committed:
Merge branch 'master' into teams

Twixes committed:
Use uuid1_macless in Person

Twixes committed:
Fix typing

Twixes committed:
Update tests

Twixes committed:
Update /api/team/signup to /api/organization/signup

Twixes committed:
Merge branch 'master' into teams

Twixes committed:
Merge branch 'master' into teams

Twixes committed:
Fix mypy issues and update tests

The remaining failures are actually missing functionality (TDD applied), so filling these in.

Twixes committed:
Update 0086_org_live

Twixes committed:
Make small improvements

Twixes committed:
Implement permissions

Twixes committed:
Remove now unneccesary membership check

Twixes committed:
Update setup_dev.py

Twixes committed:
Make small frontend improvements

Twixes committed:
Add drf-nested-routers as requirement

Twixes committed:
Remove unused import

Twixes committed:
Implemented nested routes

Twixes committed:
Solve merge conflicts

Twixes committed:
Remove cruft

Twixes committed:
Add relevant org/proj/user name to headings

Twixes committed:
Fix imports

Twixes committed:
Update migration

Twixes committed:
Replace unreliable drf-nested-routers with drf-extensions

Twixes committed:
Improve unset team handling

Twixes committed:
Make org and team creation proper

Twixes committed:
Merge branch 'master' into teams

Twixes committed:
Update migration

Twixes committed:
Merge branch 'master' into teams

Twixes committed:
Fix migration

Twixes committed:
Update TopContent

Twixes committed:
Update command palette for new sidebar structure

Twixes committed:
Merge branch 'master' into teams

Twixes committed:
Remove deprecated demo data deletion

Twixes committed:
Assume that each org has a project and fix typing

Twixes committed:
Require paid plan for multiple orgs and projects

Twixes committed:
Make HogFlix demo a separate team

Twixes committed:
Merge branch 'master' into teams

Twixes committed:
Update migration

Twixes committed:
Slightly improve style

Twixes committed:
Adjust page layout bottom padding

Twixes committed:
Make user dropdown nicer

Twixes committed:
Fix base app tests

Twixes committed:
Satisfy mypy

Twixes committed:
Fix test_leave_organization

Twixes commented:
Some notes here:

#### Testing

- `yarn build` works for me locally, but causes Docker and E2E tests to all crash here. 🤔 
- `test_feature_available_multi_tenancy` is the one `posthog` has been failing for a while and I have trouble identifying why.

Could use a fresh pair of eyes! @macobo

#### Paid functionality handling

Trying to access paid functionality now results in an "upgrade PostHog" modal:

<img width="551" alt="Screen Shot 2020-10-15 at 16 27 41" src="https://user-images.githubusercontent.com/4550621/96143477-68a69c00-0f03-11eb-832b-76f25fcc2f44.png">

This only provides the option to email us though at the moment, enhancement ideas? @paolodamico 

#### Sidebar

Since we now have the Systems Status page, I grouped it along with Licenses to sidebar option Instance (hidden on Cloud to non-staff users).

<img width="197" alt="Screen Shot 2020-10-15 at 16 29 19" src="https://user-images.githubusercontent.com/4550621/96143668-9986d100-0f03-11eb-8d70-9472c7b37f10.png">

Twixes commented:
Code review pending resolving above two issues, but I think we can go ahead with a UX review.

macobo commented:
The `yarn build` error is the following:

```
|     ERROR in ./frontend/src/scenes/sceneLogic.js
|     Module not found: Error: Can't resolve './Annotations' in '/code/frontend/src/scenes'
|      @ ./frontend/src/scenes/sceneLogic.js 120:11-122:20
|      @ ./frontend/src/scenes/App.js
|      @ ./frontend/src/index.tsx
```

Didn't go through branch history - perhaps some file got renamed and dev server never errored?

macobo commented on posthog/api/organization.py:
```diff
@@ -0,0 +1,96 @@
+    def get_object(self):
+        queryset = self.filter_queryset(self.get_queryset())
+        lookup_value = self.kwargs[self.lookup_field]
+        if lookup_value == "@current":
```
Q: Why do we have this edge case?

The frontend should already know the relevant organization id right - this seems to result in added complexity everywhere but tests?

Twixes replied:
It's less work frontend side. Though I can see a way for an inconsistency to creep in if someone changes their org in another tab for instance. 🤔 Would you suggest doing away with this?

macobo replied:
Good point re switching organizations in tabs. If possible, I think it makes sense to remove it unless we plan to make it a API-wide pattern.

Twixes replied:
Definitely planning to make use of this pattern in the `/user/` endpoint: when we remake it with DRF, let's go with `/users/@me` for accessing the currently logged in user. But in the case of `/organizations` this makes less sense.

Twixes replied:
Decided not to go forward with that approach, as the tradeoff of occasional inconsistency here is constants delays in initial page loads, as it takes a non-negligible amount of waiting until `user` is available (and only then could additional organization- or team-specific data be loaded).

macobo commented on posthog/api/organization.py:
```diff
@@ -0,0 +1,96 @@
+    def create(self, validated_data: Dict, *args: Any, **kwargs: Any) -> Organization:
+        serializers.raise_errors_on_nested_writes("create", self, validated_data)
+        request = self.context["request"]
+        with transaction.atomic():
```
Q: How is an atomic transaction different from a normal transaction?

Twixes replied:
All transactions are atomic when using Postgres with Django, but `transaction.atomic()` widens the scope of atomicity from a single `INSERT/SELECT` to all queries inside the Python context, meaning if e.g. `Team.objects.create` here fails for some reason, it will be rolled back, but also `Organization.objects.create` and `OrganizationMembership.objects.create` will be rolled back.

macobo replied:
In postgres, all INSERT/UPDATE/DELETE queries get rolled back if the wrapping transaction fails. I guess that `.atomic` in this case is just creating a normal transaction in pg at least.

Twixes committed:
Improve wording

Twixes committed:
Possibly fix import

Twixes committed:
Remove misplaced None check

Twixes committed:
Enhance org and teams APIs and add tests

Twixes committed:
Fix /api/projects for particular Team

Twixes committed:
Merge branch 'master' into teams

Twixes committed:
Improve invites and demo data

paolodamico commented on posthog/models/team.py:
```diff
@@ -106,18 +105,19 @@ class Team(models.Model):
     completed_snippet_onboarding: models.BooleanField = models.BooleanField(default=False)
     ingested_event: models.BooleanField = models.BooleanField(default=False)
     uuid: models.UUIDField = models.UUIDField(default=UUIDT, editable=False, unique=True)
+    session_recording_opt_in: models.BooleanField = models.BooleanField(default=False)
```
This seems completely unrelated to this PR and the UX for this is still in open discussion, I would remove for now.

Twixes replied:
Ah, this is actually not new at all, I just moved this line to be above the deprecated fields (as there are new such ones).

paolodamico commented on posthog/permissions.py:
```diff
@@ -0,0 +1,58 @@
+CREATE_METHODS = ["POST", "PUT"]
+
+
+def extract_organization(object: Model) -> Organization:
```
This seems like it would go in utils maybe?

Twixes replied:
It's very very permissions specific though, I feel like it'd just bloat utils when permissions is where it makes the most sense.

paolodamico commented on posthog/permissions.py:
```diff
@@ -0,0 +1,58 @@
+        return object.organization  # type: ignore
+    except AttributeError:
+        raise ValueError(
+            "Object not compatibile with organization-based permissions, as it does not have field `organization`!"
```
Minor typo

paolodamico commented on posthog/models/organization.py:
```diff
@@ -1,15 +1,33 @@
+            )
+            team = Team.objects.create(organization=organization, **(team_fields or {}))
+            user.current_organization = organization
+            user.current_team = team
```
I'm not sure about the `current_team` & `current_organization` approach of saving this stuff in the DB, seems prone to discrepancies (e.g. current team not matching current org), plus a lot of extra changes to the DB. I would consider just sending an extra header on each request indicating which team is selected on the frontend. This also makes sense if you're operating on two devices, you could easily be working with different teams on each device (e.g. I would open an incognito window if I wanted to see website numbers against app numbers).

Twixes replied:
I agree that that would be great, but it would also be a significantly larger refactor (consider insights, which would need to be project-nested then like so: `/api/projects/<project_id>/insights/` – this would have to apply to everything). Could we do this a bit later in another PR? One of the points here is that this one is already too big (which is also true).

paolodamico replied:
I was considering something easier, just add a header (e.g. `Posthog-Project`), on the backend it's a simple mixin to parse that and on the frontend I believe you can use the `api` lib that all requests use (or an interceptor). I do think this might simplify the current PR, but obviously not dead set against doing it this way, we can definitely consider this for an improvement later on.

paolodamico reviewed (changes requested):
Here are my thoughts,

### General UX
- At first glance, the project/org/user selectors were somewhat confusing, won't dive too much into this because #1775 is coming, but I think a quick improvement could be naming the first project something like "Default project" (as opposed to using the org name too), that way it's a lot more clear what's going on.

![image](https://user-images.githubusercontent.com/5864173/96411101-db539800-11df-11eb-8ceb-b02d990f2928.png)

- Accessing my settings from top bar seems to refresh the entire page instead of loading with JS.
![image](https://user-images.githubusercontent.com/5864173/96411672-c4617580-11e0-11eb-8161-8c1eea15238e.gif)

- **[FYI; non-actionable]** I'll consider this new menu layout for #1775, but I definitely think we need to try to simplify this a little bit.

- The new users inviting UX is a bit convoluted, there's too many things going on. What I'd do is just allow inviting users with an email address [_rationale_: it's a common mental model for the user (simple), it's the safest of the 3, and it removes friction for the person signing up].  You would provide the invite link to the inviting user (if the instance has SMTP capabilities send an email to the invitee automatically; N.B. tell the inviting user if the email was sent). 
![image](https://user-images.githubusercontent.com/5864173/96415179-0ccf6200-11e6-11eb-823b-d86b12084e5c.png)

- Related to the above I would not have a separate page for invites, I would just add the invite list to the top of the member list (with a visual distinction, e.g. italics or grayed out).

- **[FYI; non-actionable]** haven't tested using the invite system pending outcome on the UX points above.


### Upgrade modal
- To make it a bit more visual (plus make it easier for the user to understand what's going on if they keep running into these modals), I would consider something like below (but @lottiecoxon might be able to help with a better graphic).
![image](https://user-images.githubusercontent.com/5864173/96414409-faa0f400-11e4-11eb-80e8-1e5650593fa1.png)
- For the upgrade now button I would take the user to the relevant section of the [pricing page](https://posthog.com/pricing?o=enterprise) instead of the email, we haven't provided a lot of information at this point.
- As for the copy, maybe @yakkomajuri can provide better insight, but I would remove the "Oops!" (it's not accidental/unexpected), minor adjustment ("[...] and organizations is an advanced [..]"), keep the second line in the same paragraph and change to "Upgrade now and get access to this as well as other powerful enhancements".


As for the code review just FYI I didn't repeat comments for the same stuff (e.g. using `rest_framework.status`) in multiple occurrences.


Some overall feedback. I would strongly advise against making unrelated changes in the same PR (i.e. moving the entire frontend structure tree, or migrating a ton of unrelated stuff to TS [e.g. annotations]). While they're definitely helpful at reducing tech debt, it goes against our atomic PRs workflow, bloats PR and makes it harder for reviewing (and reverting if necessary). For context, this PR introduces more line changes than the command palette or plugins PR.

Twixes commented on ee/api/test/test_team.py:
```diff
@@ -0,0 +1,50 @@
+        self.assertEqual(response.status_code, 204)
+        self.assertEqual(Team.objects.filter(organization=self.organization).count(), 1)
+
+    def test_no_delete_last_team(self):
```
Discussed this with @timgl. Having last org/project be deletable would invalidate a lot of our current assumptions, because it's always been assumed that `User.team` is never `None`. For example in such cases most sidebar items would have to be hidden, endpoints would have to be disabled. I think we should make this flexible in the future, although I'd preface that with #1775.

paolodamico replied:
Sounds good 👌

Twixes commented on ee/api/test/test_organization.py:
```diff
@@ -0,0 +1,55 @@
+        self.assertEqual(response.status_code, 403)
+        self.assertTrue(Organization.objects.filter(id=organization.id).exists())
+        self.assertTrue(Team.objects.filter(id=team.id).exists())
+
```
That's covered by `test_organization.py` in the `posthog` (FOSS-only) app.

Twixes commented on posthog/api/organization.py:
```diff
@@ -0,0 +1,98 @@
+from posthog.permissions import CREATE_METHODS, OrganizationAdminWritePermissions, OrganizationMemberPermissions
+
+
+class PremiumMultiorganizationPermissions(permissions.BasePermission):
```
It seems like an actual permissions thing to me (especially since 402 is very rarely used) plus permission classes are very nice to use. Not sure if this is doable in a similarly simple way without permission classes.

Twixes commented on posthog/api/organization.py:
```diff
@@ -0,0 +1,98 @@
+    def get_object(self):
+        queryset = self.filter_queryset(self.get_queryset())
+        lookup_value = self.kwargs[self.lookup_field]
+        if lookup_value == "@current":
```
Basically my concern is that we'd have to run `check_object_permissions` twice then.

Twixes commented on posthog/api/organization_member.py:
```diff
@@ -0,0 +1,88 @@
+    message = "Your organization access level is insufficient."
+
+    def has_object_permission(self, request: Request, view, object: Model) -> bool:
+        if request.method in SAFE_METHODS or request.method == "DELETE":
```
What do you mean by stuff?

paolodamico replied:
So the permission validates that anyone can do any safe method (GET, HEAD, OPTIONS) but DELETE too, why allow DELETE is my question

Twixes replied:
Ah, yup, it used to be that `destroy` in the viewset itself checked if the user is only trying to delete themselves (which is allowed), not someone else (which is disallowed) – but good point, move that to this permission class so that it's purpose is clearer.

Twixes commented on posthog/api/organization.py:
```diff
@@ -0,0 +1,98 @@
+        return obj
+
+    def get_queryset(self) -> QuerySet:
+        return Organization.objects.filter(
```
I think `self.request.user.organizations.all()`, but :+1: probably a good idea.

paolodamico replied:
Right, yes!

Twixes commented on posthog/api/organization_member.py:
```diff
@@ -0,0 +1,88 @@
+    def has_object_permission(self, request: Request, view, object: Model) -> bool:
+        if request.method in SAFE_METHODS or request.method == "DELETE":
+            return True
+        organization = extract_organization(object)
```
Basically just trusting the PostHog developer putting in this check to know whether their object is organization-related.

Twixes commented on posthog/api/organization_member.py:
```diff
@@ -0,0 +1,88 @@
+class OrganizationMembershipSerializer(serializers.ModelSerializer):
+    user_first_name = serializers.CharField(source="user.first_name", read_only=True)
+    user_email = serializers.CharField(source="user.email", read_only=True)
+    membership_id = serializers.CharField(source="id", read_only=True)
```
Because the class is actually misnamed, should be `OrganizationMemberSerializer`. It's a hybrid between `User` and `OrganizationMembership`.

Twixes commented on posthog/api/organization_member.py:
```diff
@@ -0,0 +1,88 @@
+
+class OrganizationMemberViewSet(
+    NestedViewSetMixin,
+    mixins.DestroyModelMixin,
```
More verbose about intentions this way.

Twixes commented on posthog/api/organization_member.py:
```diff
@@ -0,0 +1,88 @@
+    viewsets.GenericViewSet,
+):
+    serializer_class = OrganizationMembershipSerializer
+    pagination_class = None
```
Good point. Well, handling pagination is a bit more work on the frontend, but important to keep response structure consistent.

Twixes commented on posthog/api/test/base.py:
```diff
@@ -25,38 +25,38 @@ def setUp(self):
 
 
-class APIBaseTest(APITestCase, ErrorResponsesMixin):
+class APIBaseTest(ErrorResponsesMixin, APITestCase):
```
Just consistency with the ordering in classes above.

Twixes commented on posthog/api/teams.py:
```diff
@@ -0,0 +1,178 @@
+
+    def get_queryset(self):
+        queryset = super().get_queryset()
+        if isinstance(self.request.user, AnonymousUser):
```
Yup, obviously. 👍

Twixes commented on posthog/api/organization_member.py:
```diff
@@ -0,0 +1,88 @@
+            try:
+                return queryset.filter(**parents_query_dict)
+            except ValueError:
+                raise Http404
```
Good point, this is a holdover from copying a large part of the method from the `NestedViewSetMixin` superclass (drf-extensions).

Twixes commented on posthog/api/organization_member.py:
```diff
@@ -0,0 +1,88 @@
+                parents_query_dict["organization_id"] = self.request.user.organization.id
+            try:
+                return queryset.filter(**parents_query_dict)
+            except ValueError:
```
Hm, this is how drf-exceptions handles it and it seems pretty trustworthy.  🤔

Twixes commented on posthog/api/teams.py:
```diff
@@ -0,0 +1,178 @@
+import json
```
On one hand, plural is more RESTful and I'd prefer we move to plural overall, on the other hand there'd be consistency in between. So you suggest sticking to singular?

Twixes commented on posthog/api/organization_member.py:
```diff
@@ -0,0 +1,88 @@
+        membership_to_delete = self.get_object()
+        try:
+            if (
+                membership_to_delete.user != request.user
```
Moved this to the permission class. Unsure how to standardize.

Twixes committed:
Address feedback

Twixes committed:
Merge branch 'master' into teams

Twixes committed:
Put everything related to billing on Organization

Twixes commented on ee/api/test/test_organization.py:
```diff
@@ -0,0 +1,55 @@
+    def test_no_delete_organization_not_belonging_to(self):
+        organization = Organization.objects.create(name="Some Other Org")
+        response = self.client.delete(f"/api/organizations/{organization.id}")
+        self.assertEqual(response.status_code, 403)
```
That would also make sense, although that'd require throwing the exception in a less elegant and reusable way than the permission class.

Twixes commented on posthog/api/__init__.py:
```diff
@@ -16,16 +13,21 @@
 
 
-class OptionalTrailingSlashRouter(routers.DefaultRouter):
+class HedgeRouter(ExtendedDefaultRouter):
```
True, made it less tongue-in-cheek.

Twixes commented on posthog/api/organization_member.py:
```diff
@@ -0,0 +1,88 @@
+from django.db.models import Model, QuerySet, query
```
Hmm, interesting point, could you elaborate why unnecessary?

Twixes committed:
Fix minor issues

Twixes committed:
Simplify invitation creation

Twixes committed:
Update team model

Twixes commented on ee/api/test/test_organization.py:
```diff
@@ -0,0 +1,55 @@
+
+class TestOrganizationEnterpriseAPI(APILicensedTest):
+    def test_create_organization(self):
+        response = self.client.post("/api/organizations/", {"name": "Test"})
```
Any signed in user. Now, this is actually pretty clear on self-hosted, where the license is per-instance and we can just block creating additional organizations without a license, but unsure what to do on Cloud, where billing is per-organization. So we can't really make multiple organizations a paid feature. And also can't make multiple projects then, because you can create an organization as a surrogate of that anyway. Sentry for instance, which is quite similar in how its built, puts no restrictions on org/proj creation. Thoughts?

paolodamico replied:
Well regarding on whether the feature should be restricted, I think at the very least we should start enabling it from the Starter plan, might even make sense to push to Hobby. @timgl thoughts?

As for creating orgs, what you described makes sense.

Twixes commented on frontend/src/scenes/UpgradeModal.tsx:
```diff
@@ -0,0 +1,46 @@
+            title="Unleash PostHog's Full Power"
+            okText="Upgrade Now"
+            cancelText="Maybe Later"
+            onOk={() => {
```
Fortunately since we're using Kea anyway, I added listeners with `posthog.capture`, so this is covered.

Twixes committed:
Make orgs and projects premium only on self-hosted

Twixes committed:
Improve testing

Twixes committed:
Update migration

Twixes committed:
Remove extra License import

Twixes committed:
Fix minor issues

Twixes committed:
Fix Django tests

Twixes committed:
Fix Cypress

Twixes committed:
Merge branch 'master' into teams

Twixes committed:
Fix yarn build

Twixes committed:
Fix TeamSignupViewset

Twixes committed:
Fix posthog-production incompatibility

Twixes committed:
Remove extraneous insight endpoint registration

Twixes committed:
Adjust tests for posthog-production

paolodamico reviewed (changes requested):
Some general comments (mostly UX), obviously please feel free to push back in any of them, or disregard them altogether if you think it's best:

### Org Invites
- I still think it's still a lot better to have the invites in the same page as members, it's confusing to have that separate and adds an extra unnecessary view. It also reminds the user about stale invites so they can take action (either cancel it or ping their invitee).
- I would change the "Create an invite" button to "Invite team member" or "Invite user" it's more straightforward this way.
- The ID scheme is somewhat confusing, it took me a little while to understand what was going on. The ID is also a link to do the signup but it doesn't make sense because if the inviting user clicks it, it redirects them to the insights page (on account of being logged in). My suggestion to improve this: Do not show the ID altogether, it's meaningless for the user. On the modal, after submitting the form, provide the full invite link with instructions and a copy to clipboard button. _FYI I'm creating an issue so that we can add automatic emails as a later improvement_.
- With the new invite member mental model, the "Uses so Far" and "Last Used by" don't make sense anymore. The behavior you would expect is just to see the unused invites in the list. After the invitee has signed up, the invite would automatically be deleted.
- **[Functionality]** - I was able to use a sign up link with a different email (see below).
<img width="800" alt="screenshot" src="https://user-images.githubusercontent.com/5864173/96681460-3667c500-136f-11eb-9ee9-20ef60289e9c.png">

- I don't think I should be able to delete myself (if I remember correctly this was restricted at the API level).
<img width="600" alt="image" src="https://user-images.githubusercontent.com/5864173/96681689-9c544c80-136f-11eb-8794-93d5066b3ab1.png">

- **[Functionality]** - Attempting to remove a team member made React crash (see below), but the member was indeed removed.
    - Related to the above, I was already logged in with the deleted member, I refreshed the page and got a 500 error ("Organization matching query does not exist."). I would strongly suggest just permanently deleting the user and setting foreign keys to `None` to indicate that the related objects were created/updated by a deleted user.
<img width="900" alt="image" src="https://user-images.githubusercontent.com/5864173/96682064-33b99f80-1370-11eb-9709-2e5716b002de.png">

### Creating projects & orgs
- After I created a project I was taken to the ingestion guided setup. It's a bit weird because it feels like I'm going through the initial setup again. I would suggest taking the user to the project page or something like that, and give them the option to enter the guided ingestion process independently. Definitely not force them on them because now I can't do anything (i.e. go back to my older project) until I finish this which I may not be able to do right now.
- Went through the ingestion process. The instant I got to the last step it showed me a message that I had already ingested events (without doing anything), plus after clicking "Complete", it took me back to the beginning of the ingestion process, I couldn't leave this loop at all. Same problem when creating an org.
![image](https://user-images.githubusercontent.com/5864173/96683979-c8bd9800-1372-11eb-9bf1-81aa750d324b.gif)

- Switching between projects is a bit unreliable (project is not always changed). Same for switching between orgs (see below).

![image](https://user-images.githubusercontent.com/5864173/96684687-e50e0480-1373-11eb-8668-34f86d0a0da1.gif)

Twixes committed:
Simplify invitations and fix email validation

Twixes commented on frontend/src/layout/TopContent/CommandPaletteButton.tsx:
```diff
@@ -16,7 +16,7 @@ export function CommandPaletteButton(): JSX.Element {
         >
             <SearchOutlined size={1} style={{ marginRight: '0.5rem' }} />
-            {platformCommandControlKey('K')}
+            <b>{platformCommandControlKey('K')}</b>
```
Agreed that bold 700 isn't perfect here, but regular 400 on the other hand is hardly noticeable. Made it medium 500.

Twixes commented on frontend/src/lib/components/CommandPalette/commandPaletteLogic.ts:
```diff
@@ -495,9 +489,39 @@ export const commandPaletteLogic = kea<
+                        synonyms: ['account'],
                         executor: () => {
-                            push('/team')
+                            push('/me/settings')
```
"Me" seems more distinctly personal to me than "Account". "Profile" on the other hand sounds like something public.

Twixes commented on frontend/src/scenes/organization/Invites/logic.tsx:
```diff
@@ -0,0 +1,36 @@
+                toast(
+                    <div>
+                        <h1 className="text-success">
+                            <CheckCircleOutlined /> Invite <b>{invite.id}</b> successfully removed!
```
Replaced with email.

Twixes commented on frontend/src/scenes/project/Settings/index.js:
```diff
@@ -29,46 +23,40 @@ function _Setup() {
-            These are the domains and URLs where the Toolbar will automatically open if you're logged in. It's also
-            where you'll be able to create Actions.
+            <p>
+                These are the domains and URLs where the Toolbar will automatically open if you're logged in. It's also
```
Not sure how to reword it atm, maybe add this to #1755.

Twixes commented on frontend/src/scenes/sceneLogic.js:
```diff
@@ -45,32 +49,38 @@ export const routes = {
     '/person_by_id/:id': 'person',
     '/person/*': 'person',
-    '/people': 'people',
+    '/people/persons': 'people',
```
Keeping URLs consistent with the sidebar - and Persons is part of People.

Twixes committed:
Address all feedback

Twixes committed:
Satisfy mypy

Twixes commented:
Good points, to expand on a couple:

- I'll push back on merging Members and Invites with the current settings layout, because I feel like that wouldn't work great stylistically. Keen on doing this when we reorganize the sidebar though, great point on stale invites.
  On a related note, would it make sense if the invite was simply deleted after being used up?
- Deleting yourself is basically just leaving. Though maybe administrators should not be allowed to leave, just so there never is an admin-less org. 🤔 The owner concept comes in useful here, maybe should introduce that.

Resolved all the other points.

Twixes committed:
Update migration

Twixes committed:
Fix constraint removal in migration

Twixes committed:
Update tests

Twixes committed:
Fix test creation edge case

Twixes committed:
Run posthog-production CI tests against this branch and teams-live

Twixes committed:
Merge branch 'master' into teams

Twixes committed:
Ensure that js_posthog_api_key is always passed

Twixes committed:
Fix preflight check pre-login

Twixes committed:
Update cypress tests

Twixes committed:
Update instanceStatus.js

Twixes committed:
Bring ee tets up to par

Twixes committed:
Bring actions-ux-201012 back

Twixes committed:
Merge branch 'master' into teams

Twixes committed:
Cypress retry in cypress.json

Twixes committed:
Revert "Run posthog-production CI tests against this branch and teams-live"

This reverts commit d79cb844d823d767b060122af32bd74163593d7a.
